exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,608 @@
+updates:
+	for x := range updts {
+		for tip := &ll.head; *tip != nil; tip = &(*tip).next {
+			if updts[x].ID != (*tip).value.ID {
+				continue
+			}
+			stack.Push(deleteAtTip(ll, tip), time.Now())
+			continue updates
+		}
+		if !bypassErr {
+			return fmt.Errorf("delete error: %w %d not found",
+				errIDCannotBeMatched,
+				updts[x].ID)
+		}
+	}
+	return nil
+}
+
+// cleanup reduces the max size of the depth length if exceeded. Is used after
+// updates have been applied instead of adhoc, reason being its easier to prune
+// at the end. (cant inline)
+func (ll *linkedList) cleanup(maxChainLength int, stack *stack) {
+	// Reduces the max length of total linked list chain, occurs after updates
+	// have been implemented as updates can push length out of bounds, if
+	// cleaved after that update, new update might not applied correctly.
+	n := ll.head
+	for i := 0; i < maxChainLength; i++ {
+		if n.next == nil {
+			return
+		}
+		n = n.next
+	}
+
+	// cleave reference to current node
+	if n.prev != nil {
+		n.prev.next = nil
+	} else {
+		ll.head = nil
+	}
+
+	var pruned int
+	for n != nil {
+		pruned++
+		pending := n.next
+		stack.Push(n, time.Now())
+		n = pending
+	}
+	ll.length -= pruned
+}
+
+// Amount returns total depth liquidity and value
+func (ll *linkedList) amount() (liquidity, value float64) {
+	for tip := ll.head; tip != nil; tip = tip.next {
+		liquidity += tip.value.Amount
+		value += tip.value.Amount * tip.value.Price
+	}
+	return
+}
+
+// Retrieve returns a full slice of contents from the linked list
+func (ll *linkedList) retrieve() Items {
+	depth := make(Items, ll.length)
+	iterator := 0
+	for tip := ll.head; tip != nil; tip = tip.next {
+		depth[iterator] = tip.value
+		iterator++
+	}
+	return depth
+}
+
+// bids embed a linked list to attach methods for bid depth specific
+// functionality
+type bids struct {
+	linkedList
+}
+
+// updateInsertByPrice amends, inserts, moves and cleaves length of depth by
+// updates in bid linked list
+func (ll *bids) updateInsertByPrice(updts Items, stack *stack, maxChainLength int) {
+	for x := range updts {
+		for tip := &ll.head; ; tip = &(*tip).next {
+			if *tip == nil {
+				insertHeadSpecific(&ll.linkedList, updts[x], stack)
+				break
+			}
+			if (*tip).value.Price == updts[x].Price { // Match check
+				if updts[x].Amount <= 0 { // Capture delete update
+					stack.Push(deleteAtTip(&ll.linkedList, tip), time.Now())
+				} else { // Amend current amount value
+					(*tip).value.Amount = updts[x].Amount
+				}
+				break // Continue updates
+			}
+
+			if (*tip).value.Price < updts[x].Price { // Insert
+				if updts[x].Amount > 0 { // Filter delete, should already be
+					insertAtTip(&ll.linkedList, tip, updts[x], stack)
+				}
+				break // Continue updates
+			}
+
+			if (*tip).next == nil { // Tip is at tail
+				if updts[x].Amount > 0 {
+					insertAtTail(&ll.linkedList, tip, updts[x], stack)
+				}
+				break
```

Comment:
since tip is at tail maybe we don't need the break 👀

Reply:
We don't check for nil in the for loop anymore so this will add another node.